### PR TITLE
Fix validation of ingress rules with wildcard host

### PR DESCRIPTION
**Pull request by Miciah**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Fix ingress validation so that it validates the rules of an ingress that specifies a wildcard host.  #29204 added an inopportune `continue` statement that caused this validation to be skipped.  For backwards compatibility, this change restores validation for v1 of the api but still skips it on v1beta1.

* `pkg/apis/networking/validation/validation.go` (`IngressValidationOptions`): Add `AllowInvalidWildcardHostRule` field to indicate that validation of rules should be skipped for ingresses that specify wildcard hosts.
(`ValidateIngressCreate`): Set `AllowInvalidWildcardHostRule` to true if the request is using the v1beta1 API version.
(`ValidateIngressUpdate`): Set `AllowInvalidWildcardHostRule` to true if the request or old ingress is using the v1beta1 API version.
(`validateIngressRules`): Don't skip validation of the ingress rules unless the ingress has a wildcard host and `AllowInvalidWildcardHostRule` is true.
(`allowInvalidWildcardHostRule`): New helper for `ValidateIngressCreate` and `ValidateIngressUpdate`.
* `pkg/apis/networking/validation/validation_test.go` (`TestValidateIngressCreate`, `TestValidateIngressUpdate`): Add test cases to ensure that validation is performed on v1 objects and skipped on v1beta objects for backwards compatibility.
(`TestValidateIngressTLS`): Specify `PathType` so that the test passes.

**Which issue(s) this PR fixes**:

**Does this PR introduce a user-facing change?**:

```release-note
When creating a networking.k8s.io/v1 Ingress API object, `spec.rules[*].http` values are now validated consistently when the `host` field contains a wildcard.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

----

~I'm not sure whether it is possible to fix the validation without breaking backwards compatibility.  I think `pathType` is all right because API defaulting makes sure it is set, but I am not sure what to do about ingresses that are already persisted with malformed `path` or `backend` values.~

Backwards compatibility is preserved using Jordan's patch in https://github.com/kubernetes/kubernetes/pull/93954#issuecomment-673554452.

### CODE REVIEW

**Comment by liggitt** [comment]
great catch. we'll need to ratchet this for networking.k8s.io/v1 ingress objects only with something like the following:

```diff
diff --git a/pkg/apis/networking/validation/validation.go b/pkg/apis/networking/validation/validation.go
index 526ad4f8c6f..13420d6a770 100644
--- a/pkg/apis/networking/validation/validation.go
+++ b/pkg/apis/networking/validation/validation.go
@@ -201,6 +201,9 @@ var ValidateIngressName = apimachineryvalidation.NameIsDNSSubdomain
 type IngressValidationOptions struct {
 	// AllowInvalidSecretName indicates whether spec.tls[*].secretName values that are not valid Secret names should be allowed
 	AllowInvalidSecretName bool
+
+	// AllowInvalidWildcardHostRule indicates whether invalid rule values are allowed in rules with wildcard hostnames
+	AllowInvalidWildcardHostRule bool
 }
 
 // ValidateIngress validates Ingresses on create and update.
@@ -215,7 +218,8 @@ func ValidateIngressCreate(ingress *networking.Ingress, requestGV schema.GroupVe
 	allErrs := field.ErrorList{}
 	var opts IngressValidationOptions
 	opts = IngressValidationOptions{
-		AllowInvalidSecretName: allowInvalidSecretName(requestGV, nil),
+		AllowInvalidSecretName:       allowInvalidSecretName(requestGV, nil),
+		AllowInvalidWildcardHostRule: allowInvalidWildcardHostRule(requestGV, nil),
 	}
 	allErrs = append(allErrs, validateIngress(ingress, opts, requestGV)...)
 	annotationVal, annotationIsSet := ingress.Annotations[annotationIngressClass]
@@ -231,7 +235,8 @@ func ValidateIngressUpdate(ingress, oldIngress *networking.Ingress, requestGV sc
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&ingress.ObjectMeta, &oldIngress.ObjectMeta, field.NewPath("metadata"))
 	var opts IngressValidationOptions
 	opts = IngressValidationOptions{
-		AllowInvalidSecretName: allowInvalidSecretName(requestGV, oldIngress),
+		AllowInvalidSecretName:       allowInvalidSecretName(requestGV, oldIngress),
+		AllowInvalidWildcardHostRule: allowInvalidWildcardHostRule(requestGV, oldIngress),
 	}
 
 	allErrs = append(allErrs, validateIngress(ingress, opts, requestGV)...)
@@ -313,6 +318,7 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
 		return append(allErrs, field.Required(fldPath, ""))
 	}
 	for i, ih := range ingressRules {
+		wildcardHost := false
 		if len(ih.Host) > 0 {
 			if isIP := (net.ParseIP(ih.Host) != nil); isIP {
 				allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("host"), ih.Host, "must be a DNS name, not an IP address"))
@@ -323,13 +329,17 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
 				for _, msg := range validation.IsWildcardDNS1123Subdomain(ih.Host) {
 					allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("host"), ih.Host, msg))
 				}
+				wildcardHost = true
 			} else {
 				for _, msg := range validation.IsDNS1123Subdomain(ih.Host) {
 					allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("host"), ih.Host, msg))
 				}
 			}
 		}
-		allErrs = append(allErrs, validateIngressRuleValue(&ih.IngressRuleValue, fldPath.Index(0), opts, requestGV)...)
+
+		if !wildcardHost || !opts.AllowInvalidWildcardHostRule {
+			allErrs = append(allErrs, validateIngressRuleValue(&ih.IngressRuleValue, fldPath.Index(0), opts, requestGV)...)
+		}
 	}
 	return allErrs
 }
@@ -558,3 +568,19 @@ func validateTLSSecretName(name string) []string {
 	}
 	return apivalidation.ValidateSecretName(name, false)
 }
+
+func allowInvalidWildcardHostRule(gv schema.GroupVersion, oldIngress *networking.Ingress) bool {
+	if gv == networkingv1beta1.SchemeGroupVersion || gv == extensionsv1beta1.SchemeGroupVersion {
+		// backwards compatibility with released API versions that allowed invalid rules
+		return true
+	}
+	if oldIngress != nil {
+		for _, rule := range oldIngress.Spec.Rules {
+			if strings.Contains(rule.Host, "*") && len(validateIngressRuleValue(&rule.IngressRuleValue, nil, IngressValidationOptions{}, gv)) > 0 {
+				// backwards compatibility with existing invalid data
+				return true
+			}
+		}
+	}
+	return false
+}
```

see https://github.com/kubernetes/kubernetes/pull/93929 for the style of unit test we'll want, which exercises tight validation on create and update of valid objects via networking.k8s.io/v1, and relaxed validation elsewhere

**Comment by liggitt** [comment]
/kind bug
/priority critical-urgent

to ensure correct validation for networking.k8s.io/v1 Ingress, new in 1.19

**Comment by liggitt** [comment]
looks like a rebase went sideways

**Comment by Miciah** [comment]
> looks like a rebase went sideways

Yeah, sorry, split the new test off into #93954 and pushed the wrong thing.  This PR is now focused on the change in `validateIngressRules`.

**Comment by liggitt** [comment]
let me know when https://github.com/kubernetes/kubernetes/pull/93954#issuecomment-673554452 is incorporated and we'll get this reviewed asap

**Comment by robscott** [comment]
Thanks for catching this @Miciah!

/cc @cmluciano

**Comment by Miciah** [comment]
Sorry the delay!  I have incorporated https://github.com/kubernetes/kubernetes/pull/93954#issuecomment-673554452 into the PR.  Thanks!

**Comment by Miciah** [comment]
Added unit tests for `TestValidateIngressCreate` and `TestValidateIngressUpdate`.

**Comment by robscott** [comment]
Thanks for catching this one, sorry I missed it initially! That `continue` statement skipped some very important validation...

/lgtm

File: pkg/apis/networking/validation/validation.go [hunk 1]
```diff
@@ -323,13 +329,17 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
+			} else {
+				for _, msg := range validation.IsDNS1123Subdomain(ih.Host) {
+					allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("host"), ih.Host, msg))
+				}
 			}
 		}
-		allErrs = append(allErrs, validateIngressRuleValue(&ih.IngressRuleValue, fldPath.Index(0), opts, requestGV)...)
+
+		if !wildcardHost || !opts.AllowInvalidWildcardHostRule {
+			allErrs = append(allErrs, validateIngressRuleValue(&ih.IngressRuleValue, fldPath.Index(0), opts, requestGV)...)
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go:341]: nit: preexisting, but should be `.Index(i)`

Comment (Miciah) [reply pkg/apis/networking/validation/validation.go:341]: Thanks!  Fixed.

**Pull request merged**